### DISCUSSION TITLE
Understanding errors in the context of an MQTT server

### DISCUSSION THREAD

**Discussion by ctron** [ntex-rs/ntex-mqtt#46] [Q&A] [answered]
I am struggling to understand the way errors are handled in the context of implementing an MQTT server.

In the session example, there is a limited use of `MyError`, however I don't understand what it does: https://github.com/ntex-rs/ntex-mqtt/blob/master/examples/session.rs

https://github.com/ntex-rs/ntex-mqtt/blob/f4638642d30a6f2c1b8d60ebba81ed2ea4749a31/examples/session.rs#L23-L29

I understand there are three main areas where errors generated by custom implementation code can happen (maybe more):

* During the handshake
* When receiving a request to publish
* When receiving a control message

What will happen when my code hands back `MyError` during either of those cases? And is there a way (maybe more for MQTTv5) to influence the response back to the client, for different kinds of "my custom error"? Like sending back "Access denied" when subscribing to a specific topic?

It would be great to get some help in understanding this.

**Comment by fafhrd91** [comment, marked as answer]
in v5 you can return negative ack for publish message, so TryFrom gives you this ability, you can generate whatever PublishAck you need. if you cannot handle error and need to disconnect connection
you don't convert error and return it, in that case dispatcher calls your `control` service with `ControlMessage::Error` message, and you can create proper Disconnect message

**Reply by ctron** [reply]
Hm ok … that helps me understand. Thanks.

So for v3, an error will just close the connection? And for v5 an error will get mapped by the call to "control/Disconnect", but also close the connection (with the reason code used in the `ack` call.

**Reply by fafhrd91** [reply]
v3 protocol does not support negative ack, so we can only close connection.

v5 closes connection only if error cannot be converted to PublishAck, otherwise publish ack get sent to peer
